oncall/activecalculator.go

```diff
@@ -0,0 +1,133 @@
+package oncall
+
+import (
+	"sort"
+	"sync"
+	"time"
+)
+
+var (
+	activeCalcValuePool = &sync.Pool{
+		New: func() interface{} { return make([]activeCalcValue, 0, 100) },
+	}
+)
+
+// ActiveCalculator will calculate if the current timestamp is within a span.
+type ActiveCalculator struct {
+	*TimeIterator
+
+	states []activeCalcValue
+
+	init    bool
+	active  activeCalcValue
+	changed bool
+}
+type activeCalcValue struct {
+	ID         int64
+	Value      bool
+	OriginalID int64
+}
+
+// NewActiveCalculator will create a new ActiveCalculator bound to the TimeIterator.
+func (t *TimeIterator) NewActiveCalculator() *ActiveCalculator {
+	act := &ActiveCalculator{
+		TimeIterator: t,
+		states:       activeCalcValuePool.Get().([]activeCalcValue),
+	}
+	t.Register(act)
+
+	return act
+}
+
+// Init should be called after all SetSpan calls have been completed and before Next().
+func (act *ActiveCalculator) Init() *ActiveCalculator {
+	if act.init {
+		return act
+	}
+	act.init = true
+
+	sort.Slice(act.states, func(i, j int) bool { return act.states[i].ID < act.states[j].ID })
+
+	return act
+}
+
+// SetSpan is used to set an active span.
+//
+// Care should be taken so that there is no overlap between spans, and
+// no start time should equal any end time.
+func (act *ActiveCalculator) SetSpan(start, end time.Time) {
+	if act.init {
+		panic("cannot add spans after Init")
+	}
+
+	// Skip if the span ends before the iterator start time.
+	//
+	// A zero end time indicates infinity (e.g. current shift from history).
+	if !end.After(act.Start()) && !end.IsZero() {
+		return
+	}
+
+	// Skip if the length of the span is <= 0.
+	if !start.Before(act.End()) {
+		return
+	}
+
+	act.set(start, true)
+	if !end.IsZero() {
+		act.set(end, false)
+	}
+}
+
+func (act *ActiveCalculator) set(t time.Time, isStart bool) {
+	id := t.Truncate(act.Step()).Unix()
+	originalID := id
+	if isStart && t.Before(act.Start()) {
+		id = act.Start().Unix()
+	}
+
+	act.states = append(act.states, activeCalcValue{ID: id, Value: isStart, OriginalID: originalID})
+}
+
+// Process implements the SubIterator.Process method.
+func (act *ActiveCalculator) Process(t int64) int64 {
+	if !act.init {
+		panic("Init never called")
+	}
+	if len(act.states) == 0 {
+		act.changed = false
+		return -1
+	}
+
+	v := act.states[0]
+	act.changed = v.ID == t
```

Comment:
`t` is the current timestamp.

```go
// Process implements the SubIterator.Process method.
func (act *ActiveCalculator) Process(t int64) int64 {
```

https://github.com/target/goalert/blob/temporary-schedules/oncall/timeiterator.go#L14
```go
// A SubIterator can be added to a TimeIterator via the Register method.
type SubIterator interface {
	// Process will be called with each timestamp that needs processing. Each call will be sequential, but
	// it may not be called for each step.
	//
	// Process should return the value of the next required timestamp if it is known otherwise 0.
	// If the iterator has no more events to process -1 can be returned to signal complete.
	Process(t int64) int64

	// Done will be called when the iterator is no longer needed.
	Done()
}
```

---

If it were a method meant for general use I'd have added more, but it's only purpose is to fulfill the `SubIterator` interface to be used by `TimeIterator`.

Reply:
Gotcha, I wasn't sure since the type was `int64`, I'm used to seeing `time.Time` for timestamp types. Noted 📝

Reply:
Yeah, that's why I still stuck with `t` as a bit of a hint. I did add `t` to the interface definition though to add just a bit more.